### PR TITLE
AF-1768: Errors on Windows when login user account contains special character

### PR DESCRIPTION
**Pull request by pefernan**
@tomasdavidorg @paulovmr could you please take a look?

### CODE REVIEW

**Comment by pefernan** [comment]
jenkins execute full downstream build

**Comment by pefernan** [comment]
jenkins execute full downstream build

**Comment by domhanak** [comment]
@tomasdavidorg Are you going to look at this or do you want me to take it?

**Comment by paulovmr** [comment]
jenkins execute full downstream build

**Comment by pefernan** [comment]
jenkins execute full downstream build

**Comment by domhanak** [comment]
jenkins execute full downstream build

**Comment by domhanak** [comment]
jenkins execute full downstream build

**Comment by ederign** [comment]
@paulovmr @domhanak could you please take a look at this?

**Comment by pefernan** [comment]
test failures appear to be unrelated!

**Comment by tomasdavidorg** [comment]
The test failures are related. ` | Uncaught exception: org.uberfire.java.nio.file.NotDirectoryException | {"ToSubject":"org.guvnor.common.services.shared.config.AppConfigService:RPC.loadPreferences::15:RespondTo:RPC"}`
image

**Comment by pefernan** [comment]
jenkins execucte full downstream build

**Comment by paulovmr** [comment]
jenkins execute full downstream build

**Comment by pefernan** [comment]
all green! @tomasdavidorg your turn!

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/scoped/ExperimentalStorageScope.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.uberfire.experimental.service.storage.scoped;
+
+import org.uberfire.experimental.service.definition.ExperimentalFeatureDefinition;
+
+public enum ExperimentalStorageScope {
+    GLOBAL, USER;
+
+    public static ExperimentalStorageScope getScope(ExperimentalFeatureDefinition featureDefinition) {
+        if(featureDefinition.isGlobal()) {
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/scoped/ExperimentalStorageScope.java]:
```suggestion
        if (featureDefinition.isGlobal()) {
```

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/ExperimentalFeaturesStorageImpl.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        } else {
+            settings = new Properties();
+        }
+    }
+
+    private void checkVersion() {
+        int currentVersion = Integer.parseInt(settings.getOrDefault(ExperimentalConstants.EXPERIMENTAL_VERSION_KEY, "1").toString());
+
+        if(currentVersion < ExperimentalConstants.EXPERIMENTAL_VERSION) {
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/ExperimentalFeaturesStorageImpl.java]:
```suggestion
        if (currentVersion < ExperimentalConstants.EXPERIMENTAL_VERSION) {
```

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/migration/impl/StorageMigrationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public StorageMigrationServiceImpl(Instance<StorageMigration> instance) {
+        for (StorageMigration migration : instance) {
+            migrations.add(migration);
+        }
+    }
+
+    @Override
+    public void migrate(final Integer targetVersion, final FileSystem fileSystem) {
+        migrations.iterator().forEachRemaining(migration -> {
+            if(migration.getTargetVersion() <= targetVersion) {
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/migration/impl/StorageMigrationServiceImpl.java]:
```suggestion
            if (migration.getTargetVersion() <= targetVersion) {
```

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/ExperimentalFeaturesStorageImpl.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.*;
+
+import static org.uberfire.experimental.service.storage.util.ExperimentalConstants.COMMENTS;
+
+@Named("global")
+public class ExperimentalFeaturesStorageImpl implements ExperimentalFeaturesStorage {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExperimentalFeaturesStorageImpl.class);
+
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/ExperimentalFeaturesStorageImpl.java]:
```suggestion
```

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/scoped/impl/UserExperimentalFeaturesStorageImpl.java [hunk 5]
```diff
@@ -14,41 +14,53 @@
...
+
+        checkStoragePath();
     }
 
-    @PostConstruct
-    public void init() {
-        initializeFileSystem();
+    private void checkStoragePath() {
+        Path path = fileSystem.getPath(USER_FOLDER_ROOT);
+        if(!ioService.exists(path)) {
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/scoped/impl/UserExperimentalFeaturesStorageImpl.java]:
```suggestion
        if (!ioService.exists(path)) {
```

File: uberfire-experimental/uberfire-experimental-backend/src/test/java/org/uberfire/experimental/service/BackendExperimentalFeaturesRegistryServiceImplTest.java [hunk 6]
```diff
@@ -241,31 +171,20 @@ private void testStoreFeature(String featureId, boolean newValue, ExperimentalFe
...
     }
 
-    private void init(Boolean enableExperimental, boolean loadData) {
-        if (loadData) {
-            try {
-                Path path = fileSystem.getPath(globalStorage.getStoragePath());
-                ioService.write(path, IOUtils.toString(getClass().getResourceAsStream("/test/global/regularFeatures.txt"), Charset.defaultCharset()));
+    private void init(Boolean enableExperimental) {
+
+        System.setProperty(BackendExperimentalFeaturesRegistryServiceImpl.EXPERIMENTAL_FEATURES_PROPERTY_NAME, enableExperimental.toString());
```
Comment (tomasdavidorg) [inline uberfire-experimental/uberfire-experimental-backend/src/test/java/org/uberfire/experimental/service/BackendExperimentalFeaturesRegistryServiceImplTest.java]: I would add an `@After` method which cleanup the property.

**Comment by pefernan** [comment]
@tomasdavidorg changes done!

**Comment by pefernan** [comment]
jenkins execute full downstream build please

File: uberfire-experimental/uberfire-experimental-backend/src/test/java/org/uberfire/experimental/service/BackendExperimentalFeaturesRegistryServiceImplTest.java (hunk 6 above)
Comment (pefernan) [reply uberfire-experimental/uberfire-experimental-backend/src/test/java/org/uberfire/experimental/service/BackendExperimentalFeaturesRegistryServiceImplTest.java]: oh thanks!

**Comment by pefernan** [comment]
@ederign green!

**Pull request merged**